File: shared/src/main/java/xyz/klinker/messenger/shared/widget/MessengerAppWidgetProvider.java
/*
 * Copyright (C) 2016 Jacob Klinker
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.klinker.messenger.shared.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.util.Log;
import android.widget.RemoteViews;

import xyz.klinker.messenger.shared.R;
import xyz.klinker.messenger.shared.MessengerActivityExtras;
import xyz.klinker.messenger.shared.data.Settings;
import xyz.klinker.messenger.shared.util.ActivityUtils;
import xyz.klinker.messenger.shared.util.ImageUtils;

public class MessengerAppWidgetProvider extends AppWidgetProvider {

    private static final String TAG = "AppWidgetProvider";

    public static final String REFRESH_ACTION = "xyz.klinker.messenger.shared.widget.REFRESH";
    public static final String OPEN_ACTION = "xyz.klinker.messenger.shared.widget.OPEN";
    public static final String EXTRA_ITEM_ID = "xyz.klinker.messenger.shared.widget.EXTRA_ITEM_ID";

    @Override
    public void onReceive(Context context, Intent intent) {
        superOnReceive(context, intent);

        AppWidgetManager mgr = getAppWidgetManager(context);

        String action = intent.getAction();
        if (action != null && action.equals(OPEN_ACTION)) {
            long itemId = intent.getLongExtra(EXTRA_ITEM_ID, 0);
            Intent openConversation = ActivityUtils.buildForComponent(ActivityUtils.MESSENGER_ACTIVITY);
            openConversation.putExtra(MessengerActivityExtras.EXTRA_CONVERSATION_ID, itemId);
            openConversation.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            context.startActivity(openConversation);
            return;
        }

        ComponentName thisAppWidget = new ComponentName(context.getPackageName(), MessengerAppWidgetProvider.class.getName());
        int[] appWidgetIds = mgr.getAppWidgetIds(thisAppWidget);

        try {
            mgr.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.widget_list);
        } catch (NullPointerException e) {
            Log.e(TAG, "failed to notify of widget changed", e);
        }
    }

    public AppWidgetManager getAppWidgetManager(Context context) {
        return AppWidgetManager.getInstance(context);
    }

    public void superOnReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
    }

    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int i = 0; i < appWidgetIds.length; ++i) {
            Intent intent = new Intent(context, MessengerWidgetService.class);
            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[i]);
            intent.setData(Uri.parse(intent.toUri(Intent.URI_INTENT_SCHEME)));

            Intent compose = ActivityUtils.buildForComponent(ActivityUtils.COMPOSE_ACTIVITY);
            PendingIntent pendingCompose = PendingIntent.getActivity(context, 0, compose, 0);

            Intent open = ActivityUtils.buildForComponent(ActivityUtils.MESSENGER_ACTIVITY);
            open.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            PendingIntent pendingOpen = PendingIntent.getActivity(context, 0, open, 0);

            RemoteViews rv = new RemoteViews(context.getPackageName(), R.layout.appwidget);
            rv.setRemoteAdapter(R.id.widget_list, intent);
            rv.setOnClickPendingIntent(R.id.compose, pendingCompose);
            rv.setOnClickPendingIntent(R.id.title, pendingOpen);

            rv.setEmptyView(R.id.widget_list, R.id.widget_empty);

            Settings settings = Settings.get(context);
            if (settings.useGlobalThemeColor) {
                Bitmap color = ImageUtils.createColoredBitmap(settings.globalColorSet.color);
                rv.setImageViewBitmap(R.id.toolbar, color);
            }

            Intent openIntent = new Intent(context, MessengerAppWidgetProvider.class);
            openIntent.setAction(MessengerAppWidgetProvider.OPEN_ACTION);
            openIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[i]);
            intent.setData(Uri.parse(intent.toUri(Intent.URI_INTENT_SCHEME)));
            PendingIntent openPendingIntent = PendingIntent.getBroadcast(context, 0, openIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT);
            rv.setPendingIntentTemplate(R.id.widget_list, openPendingIntent);

            appWidgetManager.updateAppWidget(appWidgetIds[i], rv);
        }

        super.onUpdate(context, appWidgetManager, appWidgetIds);
    }

    public static void refreshWidget(Context context) {
        int[] ids = AppWidgetManager.getInstance(context)
                .getAppWidgetIds(new ComponentName(context, MessengerAppWidgetProvider.class));

        Intent intent = new Intent(context, MessengerAppWidgetProvider.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);

        context.sendBroadcast(intent);
        context.sendBroadcast(new Intent(REFRESH_ACTION)); // send to dashclock
    }

}

File: shared/src/main/java/xyz/klinker/messenger/shared/widget/MessengerAppWidgetProvider.kt
/*
 * Copyright (C) 2017 Luke Klinker
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.klinker.messenger.shared.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.net.Uri
import android.util.Log
import android.widget.RemoteViews

import xyz.klinker.messenger.shared.R
import xyz.klinker.messenger.shared.MessengerActivityExtras
import xyz.klinker.messenger.shared.data.FeatureFlags
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.data.pojo.BaseTheme
import xyz.klinker.messenger.shared.util.ActivityUtils
import xyz.klinker.messenger.shared.util.ImageUtils

class MessengerAppWidgetProvider : AppWidgetProvider() {

    override fun onReceive(context: Context, intent: Intent) {
        superOnReceive(context, intent)

        val mgr = getAppWidgetManager(context)

        val action = intent.action
        if (action != null && action == OPEN_ACTION) {
            val itemId = intent.getLongExtra(EXTRA_ITEM_ID, 0)
            val openConversation = ActivityUtils.buildForComponent(ActivityUtils.MESSENGER_ACTIVITY)
            openConversation.putExtra(MessengerActivityExtras.EXTRA_CONVERSATION_ID, itemId)
            openConversation.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            context.startActivity(openConversation)
            return
        }

        val thisAppWidget = ComponentName(context.packageName, MessengerAppWidgetProvider::class.java.name)
        val appWidgetIds = mgr.getAppWidgetIds(thisAppWidget)

        try {
            mgr.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.widget_list)
        } catch (e: NullPointerException) {
            Log.e(TAG, "failed to notify of widget changed", e)
        }

    }

    fun getAppWidgetManager(context: Context): AppWidgetManager {
        return AppWidgetManager.getInstance(context)
    }

    fun superOnReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        for (i in appWidgetIds.indices) {
            val intent = Intent(context, MessengerWidgetService::class.java)
            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[i])
            intent.data = Uri.parse(intent.toUri(Intent.URI_INTENT_SCHEME))

            val compose = ActivityUtils.buildForComponent(ActivityUtils.COMPOSE_ACTIVITY)
            val pendingCompose = PendingIntent.getActivity(context, 0, compose, 0)

            val open = ActivityUtils.buildForComponent(ActivityUtils.MESSENGER_ACTIVITY)
            open.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            val pendingOpen = PendingIntent.getActivity(context, 0, open, 0)

            val rv = RemoteViews(context.packageName, R.layout.appwidget)
            rv.setRemoteAdapter(R.id.widget_list, intent)
            rv.setOnClickPendingIntent(R.id.compose, pendingCompose)
            rv.setOnClickPendingIntent(R.id.title, pendingOpen)

            rv.setEmptyView(R.id.widget_list, R.id.widget_empty)

            when(Settings.baseTheme) {
                BaseTheme.ALWAYS_DARK -> rv.setInt(R.id.widget_list, "setBackgroundColor", Color.parseColor("#202B30"))
                BaseTheme.BLACK -> rv.setInt(R.id.widget_list, "setBackgroundColor", Color.BLACK)
                else -> rv.setInt(R.id.widget_list, "setBackgroundColor", context.resources.getColor(R.color.background))
            }

            baseWidgetTheme = Settings.baseTheme

            val color = ImageUtils.createColoredBitmap(Settings.mainColorSet.color)
            rv.setImageViewBitmap(R.id.toolbar, color)

            val openIntent = Intent(context, MessengerAppWidgetProvider::class.java)
            openIntent.action = MessengerAppWidgetProvider.OPEN_ACTION
            openIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[i])
            intent.data = Uri.parse(intent.toUri(Intent.URI_INTENT_SCHEME))
            val openPendingIntent = PendingIntent.getBroadcast(context, 0, openIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT)
            rv.setPendingIntentTemplate(R.id.widget_list, openPendingIntent)

            appWidgetManager.updateAppWidget(appWidgetIds[i], rv)
        }

        super.onUpdate(context, appWidgetManager, appWidgetIds)
    }

    companion object {
        private val TAG = "AppWidgetProvider"

        const val REFRESH_ACTION = "xyz.klinker.messenger.shared.widget.REFRESH"
        const val OPEN_ACTION = "xyz.klinker.messenger.shared.widget.OPEN"
        const val EXTRA_ITEM_ID = "xyz.klinker.messenger.shared.widget.EXTRA_ITEM_ID"

        var baseWidgetTheme = BaseTheme.DAY_NIGHT

        fun refreshWidget(context: Context) {
            try {
                val ids = AppWidgetManager.getInstance(context)
                        .getAppWidgetIds(ComponentName(context, MessengerAppWidgetProvider::class.java))

                val intent = Intent(context, MessengerAppWidgetProvider::class.java)
                intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
                intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)

                context.sendBroadcast(intent)
                context.sendBroadcast(Intent(REFRESH_ACTION)) // send to dashclock
            } catch (e: Exception) {
            }
        }
    }

}
